### ISSUE TITLE
Support `x-goog-api-client` in go

### ISSUE THREAD

**Issue by saicheems** [googleapis/toolkit#35]
Using this document to get opinions before a PR:

Background
=========
We plan to support an `X-Goog-Api-Client` header that allows us to track the version information of the code generator and GAX libraries for each client call.

The format is `<clientlib>/<semver> gapic/<semver> gax/<semver> language/<semver>`
For example, in python: `gcloud-python/<gcloud-python-version> gapic/0.1.0 gax/0.5.0 python/2.7`

Proposal
=======
In order to support the header, we need to be able to pass optional `app_name` and `app_version` strings to the client struct on initialization.

The method signature for `NewClient` is currently:
`func NewClient(ctx context.Context, opts ...cloud.ClientOption) (*Client, error)`

We propose instead:
`func NewClient(ctx context.Context, opts ...gax.ClientOption) (*Client, error)`
where the interface `gax.ClientOption` is provided by the following:

within `gax-golang`
-> `WithCloudClientOption(cloud.ClientOption) ClientOption`
-> `WithAppVersion(appVersion string) ClientOption`
-> `WithAppName(appName string) ClientOption`

This design also allows us to implement client level options specific to the gapic surface:
-> `WithCallOptionsOverride(map[string][]gax.CallOption) ClientOption`
-> `WithTimeout(duration time.Duration) ClientOption`

Now, within `NewClient` we can generate the header string as a private global variable and provide it to all calls via `grpc.CallOption`.

GAX will also support a global `Version string` that contains a semantic version tag. It's a workaround for go's package versioning limitations, and not perfect, but hopefully sufficient for now.
@tbetbetbe's suggestion was to use a script to generate a go file with a version tag (maybe the commit hash, or github release version).

**Comment by saicheems** [comment]
Please let me know if this seems reasonable.

@vchudnov-g 
@pongad

**Comment by pongad** [comment]
I have a few thoughts:
- I would like to avoid adding options into more places. I think there are enough "entry points" as it is.
- From the look of it, we can make use of gax.callSettings and add a metadata.MD field to it. We can inject this the same way we inject defaultRetryParams(). 
- gax.Invoke can inspect the MD field. If not nil, call grpc.SendHeader.

**Comment by saicheems** [comment]
We still need to have `AppName` and `AppVersion` be set at the client level, so that has to be passed on initialization... We could have a `type ClientSettings struct` that contains those values, but I preferred the `CallOption` approach to that.

There are also other possible client level options that we can't use `cloud.ClientOption` to support.

**Comment by pongad** [comment]
I think we are misunderstanding each other. I fully support using CallOption to set the value, but the value needs to live somewhere. Since CallOption modifies callSettings, the value needs to live in callSettings.

**Comment by saicheems** [comment]
After significant offline discussion:

We decided the best approach might be to override the `cloud.ClientOption` and provide our own `gax.ClientOption` with the same + our own functionality.

This gives us the added benefit of not requiring the user to import `cloud` to use the generated surface.

We also realized that the current codegen relies on an internal package within gcloud... so the proposed solution is to copy that function `transport.DialGRPC` into GAX and implement it using the `gax.ClientOption` interface instead.
